### ISSUE TITLE
Regex matching

### ISSUE THREAD

**Issue by pokrakam** [exercism/snippet-extractor#41]
Suggestion: Could regex be incorporated in the matcher? 

One immediate use case in ABAP is an end of line comment `"` is OK and `"` in strings need to be kept, but a line starting with `"` or `*` should be skipped.

```ABAP
* Now we will loop through the table
loop at sometable reference into data(row).
  if row->somefield = 'foo'. 
    "Exclude these
    continue.
  endif.
  write / |Row { sy-tabix } contains "{ row->anotherfield }"|. 
endloop. "end-of line comments can be deleted or not...
```

Should go to
```ABAP
loop at sometable reference into data(row).
  if row->somefield = 'foo'. 
    continue.
  endif.
  write / |Row { sy-tabix } contains "{ row->anotherfield }"|. 
endloop. "end-of line comments can be deleted or not...
```

Easy to do with regex, but I can't see a way according to the current documentation.

**Comment by iHiD** [comment]
I think you can do this with the basic mode syntax: https://github.com/exercism/snippet-extractor/blob/main/docs/basic.md

But if not, you can definitely do it with extended mode: https://github.com/exercism/snippet-extractor/blob/main/docs/extended.md

@joshiraez Could you lend a hand pls :)

**Comment by pokrakam** [comment]
Thanks, I'm not sure this is the case. The extended mode syntax itself is not compatible with regex. I can enter `*` in extended mode to match a `*` and nothing else, whereas in regex it would match anything. I did try matching `^"` but it did not work.

**Comment by joshiraez** [comment]
I think you are confused on what extended mode does.

It's a design decision to not use regex and I dont foresee it changing because of two reasons

- UX: a lot of people hate regex and it gets cumbersome for very basic cases. Even very simple checks can quickly go awry when you try to do global searches
- Efficiency: regex starts to be REALLY slow the moment you introduce lookouts. For something that was going to run over a lot of files, regex seemed like it could be a pipe factor

I understand that for regex users, changing to this syntsx can be frustrating, specially because it brings a subset of regex features and it can be hard to adjust, but it has its reasons

If you are interested, it uses a trie parser behind the doors using as basis the syntax that the snippet refactor used to have.

Now for your problem. You'd need the extended syntax (so it looks up the whole file) but it  would suffice with two lines

`
!e
*
"
`

Those 2 lines without modifiers mean "any line that starts with * or " must be skipped". It ignores tabs, so don't worry about that.

In regex, they are similar to

[\*\"].*n

I think this suffices what you are asking for. As those in the end of line shouldnt be skipped, dont they?

In any case, please try it and we can iterate from there

**Comment by joshiraez** [comment]
I'd appreciate if you could give me an example with all the cases you need
to be removed and I'll be happy to work on it and explain it later to you
in case you need anything more.

If you can send me a original/extractwd to build an unit test and wprk from
there it'd be amazing

**Comment by joshiraez** [comment]
Btw, I dont remember if i did it in the end but i had the idea you could do something like

Class -->> definition. -->> endclass

If thats not implemented ill have to extend the extractor a little, but it should be no biggie.

**Comment by pokrakam** [comment]
Thanks, I did try the chained variant and it didn't work. I even tried a real dodgy hack:
```
class-->>defin\pj
ition.-->>endclass\p
```
Not sure why that won't work, but it would fall apart if "defin" is part of a class name.

**Comment by joshiraez** [comment]
Because it stops at the first match

Ill get on with implementing the multi arrow then. Or see any way to
support this.

**Comment by SleeplessByte** [comment]
@joshiraez can we introduce ~ as "includes" ? then you could do `~ implementation`

**Comment by joshiraez** [comment]
Mmm, can you put how the rule would be written with your syntax? I dont
think the double arrow has good ux, but from what you said
`class.~definition.~end.` seems even more confusing if you have to remember
what differences are between arrow and this one. Meanwhile,
`class.-->>definition.-->>end.` seems self explanatory.

In any case, is not an easy a fix as I thought. Im going to spend a few
days to do a small refactor on the trie parser. We can discuss the syntax
here in the meanwhile.

**Comment by pokrakam** [comment]
I think a wildcard would be useful. Many apps that don't use regex still use `*` or `?` as wildcards. I see `*` is already in use (and `~` for that matter), but wouldn't be a big job to escape the existing ones - happy to do it. 
So in my example it would be:
`class*definition-->>endclass`
This seems intuitive to me.

**Comment by joshiraez** [comment]
But in your example the wildcard is exactly the same as the arrow.

Also, if you use * as wildcard you are then losing that character as a
valid input character, when its one of the most common ways to comment
stuff.

I dont think we should copy regex syntax as I said before, just because its
regex syntax

**Comment by joshiraez** [comment]
I mean, if you are so keen on introducing regex, I can just add an argument
to the extended mode to use regex instead of the trie parsers. I just dont
think is a good design solution and it was one of the things we wanted to
avoid from the start, but I feel like thats what you are pushing for...?

What does everyone think about this?

**Comment by joshiraez** [comment]
And of course, if a regex mode is introduced, we wouldnt be providing
support to figure out which regex you have to build, as its what we wanted
to avoid from the start.

If you need wildcards you can use the arrow. But what you are really
looking for is how to check for a series of keywords to skip the whole
section. Im working on that now.

**Comment by pokrakam** [comment]
I think you misunderstood, I meant that * could be treated as a match only on a single line, similar to the current `+` operator. 
`-->>` is clearly documented as multi-line matching and is a different feature to my mind.

Doesn't have to be a full regex mode. I don't know trier, but it just feels to me that `*` wildcard is everyday syntax many people would already be used to, e.g. editing `.gitignore` or CLI stuff like `find . -name "abc*.adoc"`.

As to `*` already being used, if it can be escaped with `\*` then it's not a big job to replace it in the existing extractors - I've already volunteered 🙂

Of course if this is a pig to implement then we can go with whatever works.

**Comment by joshiraez** [comment]
Not a pig... I just think that using * because * is used for regex
searching in texts where * is not a common character, for a parser where we
will looking for mostly * and will need to be escaped in almost every
language (on top of that, I think we dont have anything implemented for
escaping characters) is a really bad idea.

On top of that, adding it is "hard".

The + sign exists because there were a few languages that really required
to parse up to the end of the repeated character. I think it was with the
HEREDOC syntax where you could have multiple brackets. And it was really
hacky because, well, we usually want to look for language keywords, that
are quite well defined. The trie parser can look for them, but is not the
usual neither the ideal scenario because you have to loop in place until
you find some other character. And then you get to the world of "which rule
has preference when a + is there".

If you think on any language that really requires * because you have
keywords that have variable number of variable characters in it, yeah, ok.
But so far, for what we are looking for, and how languages have to be
designed to be parsed efficiently, it really doesn't seem like a necessity
but more of a "i want this because i feel familiar with it". And, as I
said, seeing the complexity it could generate both for devs and users, Id
have to push heavily against it.

In fact, that kind of wildcard matching would break the refactor im working
on open, because now Ill be parsing keywords/rules instead of parsing using
a read and act pointer. Well, not break, because we are supporting the +
and supporting the * would be extending that, but the number of edge cases
rises exponentially and, honestly, I fail to see any example in a language
where you would need to skip 0..n characters in a keyword. If you know some
more, please be so kind as to share them to make sure we support them.

All that said, Ill go forward with enabling a regex mode where youll be
able to use the features you are expecting from regex using a regex, as I
foresee more people wanting the familiarity of regex instead of using the
extended mode

**Comment by pokrakam** [comment]
> On top of that, adding it is "hard".

That's what I meant with "a pig to implement". Our biggest real-world need is to skip a section to strip the identical bits, the rest is all nice to have / discussion topic. So if a chained `a-->>b-->>c` works then I'm happy with that. 
I had assumed that '\' escaping was already a feature because there's some mention of escaping in the docs, but if that messes things up then just go with whatever is the next best step.

**Comment by pokrakam** [comment]
@iHiD the most critical issue is that all ABAP snippets for an exercise look the same (screenshot above).

ABAP is a strongly typed language, class definitions are akin to a type declaration and coded separate from implementation.
 
We provide the definition part for our exercise templates, so these are usually the same for all exercises. I'd like to strip this, but since the keywords `class ... endclass.` describe both, the current parser can only recognise them both or not at all.

```abap
class myclass definition. 
  public section.
    "! AbapDoc comment 
    methods foo.
    methods bar returning value(result) type string.
  private section.
    data stuff type string.
endclass.                    

class myclass implementation.

  method foo.
    " Some comment
    stuff = 'Hello world!'.
  endmethod.

  method bar.
    result = stuff.
  endmethod.

endclass.
```

Desired result:
```abap
class myclass implementation.

  method foo.
    " Some comment
    stuff = 'Hello world!'.
  endmethod.

  method bar.
    result = stuff.
  endmethod.

endclass.
```

Nice to have: 
```abap
  method foo.
    stuff = 'Hello world!'.
  endmethod.

  method bar.
    result = stuff.
  endmethod.
```

**Comment by iHiD** [comment]
@pokrakam Thank you.

@joshiraez So let's focus on getting that first usecase done via the extended syntax for now? And then we can explore more options around the "Nice to have" from there?

(cc @ErikSchierboom for visibility too)

**Comment by joshiraez** [comment]
I'm on it :)

**Comment by joshiraez** [comment]
Small heads up that Im still on it. I got into a new job recently so I have been a bit absent this past months.

In the end it will require a refactor, but I think it should make it much easier to maintain. Ill keeo you posted

**Comment by larshp** [comment]
@joshiraez any updates?

**Comment by joshiraez** [comment]
Yeah, I'm really sorry. I'm really busy with work and personal stuff. I have a solution drafted but having to jump into ruby again is making it really difficult.

I still have it in my mind, I'll try to get into this as soon as I have some holidays/free time. My apologies :(
